broadinstitute/single_cell_portal_core
# Enabling Sentry Replays, silence Bard errors (SCP-5027, SCP-5172)

#### BACKGROUND & CHANGES
This update enables Sentry [session replays](https://docs.sentry.io/product/session-replay/) to allow developers better insight into user behavior leading up to an exception reported to Sentry.  This is useful in determining what actions led up to the error.  Breadcrumbs are the only user log that we have currently, and are mostly (but not universally) helpful.  Now, we will have a short video showing the behavior of the user.  This is being done on a trial basis to determine the overall efficacy of using replays in triaging bugs/issues.

Example replay: https://broad-institute.sentry.io/replays/single-cell-portal:60ff05cd733a4108adb20d4cd31b5cf3/?project=1424198&query=&referrer=%2Freplays%2F&statsPeriod=14d&yAxis=count%28%29

Additionally, this update will also permanently silence reporting of upstream Bard errors.  These errors currently make up more than half of all error reports (almost [4K per month](https://broad-institute.sentry.io/issues/?environment=production&project=1424198&query=is%3Aunresolved+%21level%3Ainfo+logger%3Aapp-frontend+message%3A%22when+logging+event+to+Bard%22&referrer=issue-list&sort=freq&statsPeriod=30d)), and are both unhelpful and not actionable.  Attempts to silence them in Sentry have been unsuccessful, so we are now opting to prevent reporting them altogether.

Some notes on current drawbacks for replays:
- Our plan only allows for 500 replays per month, which will not be sufficient if we determine that we want to leverage replays moving forward
- I have set sampling at only 5% of all errors to help deal with this so that we don't use up all of our replays immediately, meaning we may not get a replay on an error we want
- SVG and Canvas elements are not sent along with replays, meaning we will not see visualizations

#### MANUAL TESTING
1. Boot as normal and load any study with all visualizations enabled
2. In `app/javascript/lib/sentry-logging.js`, edit `getIsSuppressedEnv()` to return `false`:
```
function getIsSuppressedEnv() {
  const env = getSCPContext().environment
  // Return `false` if manually locally testing Sentry logging
  return false // ['development', 'test'].includes(env)
}
```
3. In `setUpSentry()`, edit `replaysOnErrorSampleRate` on line 126 t0 `1.0` to ensure your error is sampled
4. Back in SCP, run a gene search for 2 genes to load the correlation plot
5. Click on the legend in the correlation plot to trigger an exception
6. Go to the [replays console](https://broad-institute.sentry.io/replays/?project=1424198) and locate your error 
7. To test that we are ignoring Bard errors, sign in with any account that you know is not registered in the dev/alpha instance of Bard, such as your personal Gmail account
8. Reload the page, open the JS console and look for the following message:
```
POST https://terra-bard-dev.appspot.com/api/event/ 503
Suppressing error report to Sentry: this error is permanently ignored
```
